rosidl_runtime_rs/src/string.rs

```diff
@@ -224,6 +225,59 @@ macro_rules! string_impl {
             }
         }
 
+        impl FromIterator<char> for $string {
+            fn from_iter<I: IntoIterator<Item = char>>(iter: I) -> Self {
+                let mut buf = <$string>::default();
+                buf.extend(iter);
+                buf
+            }
+        }
+
+        impl<'a> FromIterator<&'a char> for $string {
+            fn from_iter<I: IntoIterator<Item = &'a char>>(iter: I) -> Self {
+                let mut buf = <$string>::default();
+                buf.extend(iter);
+                buf
+            }
+        }
+
+        impl Extend<char> for $string {
+            fn extend<I: IntoIterator<Item = char>>(&mut self, iter: I) {
+                let mut v = self.to_vec();
+                let iterator = iter.into_iter();
+
+                iterator.for_each(|c: char| {
+                    // UTF-8 and UTF-16 encoding requires at least 4 bytes for any character.
+                    // Technically UTF-16 could just use a buffer size of 2 here, but that's more
+                    // trouble than its worth since we are defining this function in the macro.
+                    // See https://doc.rust-lang.org/std/primitive.char.html#method.encode_utf8
+                    let mut buf = [0; 4];
+                    c.$encoding_func(&mut buf);
+
+                    let filtered_bytes: Vec<$unsigned_char_type> = buf
+                        .into_iter()
+                        .filter(|&c| c != (0 as $unsigned_char_type))
+                        .collect();
+
+                    for encoded_char in filtered_bytes {
+                        v.push(encoded_char as $char_type);
+                    }
+                });
```

Comment:
You could probably simplify this to something like 
```
let mut s = self.to_string();
s.extend(iter);
*self = Self::from(&s);
```

This should be more performant too (partly because `String::extend()` reserves an appropriate amount of space upfront if that information is available from the iterator's `size_hint()` function). It would be possible to get even more efficient and eliminate the back-and-forth conversion, but that requires dropping into some quite low-level code, so I'm also fine with not doing that. I'll leave it up to you whether you're interested in that or not.

Reply:
Ah yeah this is better. Simplified implementation, a few extra ergonomic trait impls for the String types, and as you mentioned likely more performant. Thanks for the help :+1: